clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -206,11 +209,21 @@ public double compressionRate() {
     }
 
     public boolean isFull() {
-        return recordsBuilder.isFull();
+        return !isWritable || recordsBuilder.isFull();
+    }
+
+    public void closeWithTransactionState(long pid, short epoch, int baseSequence) {
+        recordsBuilder.closeWithTransactionState(pid, epoch, baseSequence);
+        isWritable = false;
```

Comment:
The problem is that we write the headers (sequence number, pid, etc.), on close. So the only time we can do a successful close (ie when we want to actually send the batch), is at the moment before we send it (ie. in `drain`). That is why this new variable was introduced: we can't simply close the batch when there is no further room, etc. because then we would have to rewrite the headers later if we actually wanted to send the batch. 

The choice is between this variable (which blocks further writes _without_ closing), and the rewriting the final record everytime 'close' is called.. But the latter will _always_ happen on a successful send, so this seems like the better approach.